src/radical/pilot/utils/serializer.py

```diff
@@ -129,17 +129,25 @@ def deserialize_file(fname):
 #
 def deserialize_obj(obj):
     '''
-    Deserialize object from str
+    Deserialize object from str.
+
+    Raises
+    ------
+    TypeError
+        if object cannot be deserialized.
+    ValueError
+        if no result is found.
+
     '''
     result = None
 
     try:
         result = dill.loads(obj)
     except Exception as e:
-        raise Exception ("failed to deserialize from object") from e
+        raise TypeError("failed to deserialize from object") from e
```

Comment:
Do we really know that this is caused by an type error?
```py
>>> dill.loads(b'x')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/mnt/home/merzky/radical/radical.pilot.devel/ve3/lib/python3.6/site-packages/dill/_dill.py", line 327, in loads
    return load(file, ignore, **kwds)
  File "/mnt/home/merzky/radical/radical.pilot.devel/ve3/lib/python3.6/site-packages/dill/_dill.py", line 313, in load
    return Unpickler(file, ignore=ignore, **kwds).load()
  File "/mnt/home/merzky/radical/radical.pilot.devel/ve3/lib/python3.6/site-packages/dill/_dill.py", line 525, in load
    obj = StockUnpickler.load(self)
_pickle.UnpicklingError: invalid load key, 'x'.
```

Reply:
same as above.

Reply:
TypeError or ValueError both seem reasonably appropriate, since either could mean that Python objects with incompatible interfaces were asked to interact.

In this case, it is a recoverable condition, so RuntimeError seems a bit strong. Also, RuntimeError can be triggered by other parts of this use case, and I really don't want to make it harder to catch and discriminate errors specifically related to the deserialization attempt.

Here, we already see that dill may raise exceptions that don't inherit from `dill.PickleError`, and I don't think we can assume that they will otherwise derive from `pickle.PickleError`.